tickets/DM-22137

fred3m opened the pull request:
Integrate scarlet 1.0 into the stack and test on HSC data.

Commit:
Implement ADAM (scarlet 1.0)

Commit:
Allow the normalization to be modified

In a conversation with @pmelchior it was suggested that normalizing
the morphology might not be necessary with the new ADAM algorithm in
scarlet, since it's better at breaking up degeneracies.
In order to test this on HSC data this commit added functionality
to rmove the normalization constraint. The result of the test was
that deblending takes longer with poorer results when not using
normalization, so this code will be stripped out in the next commit
but is stored here for historical purposes.

Commit:
Revert "Allow the normalization to be modified"

This reverts commit 21220153b51aed447273914b045d22325e258d66.

Commit:
Change default max iterations to 300

Testing has shown that using the lower convergence threshold for
scarlet 1.0 improves the residuals but takes more iterations.
Execution on a few small 1k x 1k patches running for up to 1000
iterations showed that most models converge before 100 iterations
but a significant fraction of blends don't converge until between
200-300 iterations, while blends that haven't converged by 300
iterations do not appear to ever converge. So we choose 300
iterations as the new cutoff.

Commit:
Update tests

This updates the unit tests to use the new API.
This includes removing a few tests for functionality that has either
been stripped or significantly modified.

taranu commented on python/lsst/meas/extensions/scarlet/deblend.py:
```diff
@@ -135,26 +141,29 @@ def deblend(mExposure, footprint, log, config):
-        for center in footprint.peaks
-    ]
+    psfShape = (config.modelPsfSize, config.modelPsfSize)
+    model_psf = PSF(partial(gaussian, sigma=.8), shape=(None,)+psfShape)
```
Isn't that sigma=.8 important enough to set in the config with that default?

taranu commented on python/lsst/meas/extensions/scarlet/deblend.py:
```diff
@@ -184,20 +197,25 @@ class ScarletDeblendConfig(pexConfig.Config):
              "PSF in each band or use the same morphology in all bands"))
+    modelPsfSize = pexConfig.Field(
+        dtype=int, default=11,
+        doc="The model PSF will be (modelPsfSize, modelPsfSize)")
```
"Model PSF side length in pixels" or something like that?

taranu commented on python/lsst/meas/extensions/scarlet/deblend.py:
```diff
@@ -330,6 +348,13 @@ def _addSchemaKeys(self, schema):
+                                                            "of the source converged, otherwise it"
+                                                            "contains the bit flag for parameters that"
+                                                            "failed, which might differ depending on the"
+                                                            "source type")
```
I find it somewhat confusing that deblend_blendConvergenceFailed is a flag whose name doesn't end in _flag (why not deblend_blendConvergence_flag?), whereas deblend_sourceConvergenceFlag isn't actually a flag (how about something like deblend_sourceConvergence_bits?).

fred3m replied:
`deblend_blendConvergenceFailed` -> `deblend_blendConvergenceFailedFlag`
`deblend_sourceConvergenceFlag` -> `deblend_sourceConvergenceBitFlag`

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+    it finds a model that can be initialized. If all of the
+    models fail, including a `PointSource` model, then skip
+    the source.
+
```
I realize one can and should refer to scarlet's source here but it would be helpful to give a brief hint/summary as to how/why source initialization might fail.

fred3m replied:
clarified the docs

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+                                          monotonic=monotonic, thresh=thresh)
+            if (np.any([np.isnan(c.sed) for c in components]) or
+                    np.all([c.sed <= 0 for c in source.components])):
+                logger.warning("Could not initialize")
```
This warning isn't very helpful. Why isn't this caught upstream?

fred3m replied:
This was part of a discussion that went on for weeks. scarlet used to throw a `SourceInitError` to inform the user that initialization failed but Peter felt strongly about having users check for a bad initialization and I finally got tired of fighting him on it.

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+        except Exception:
+            # If the MultiComponentSource failed to initialize
+            # try an ExtendedSource
+            components -= 1
```
Are none of the Exceptions raised by scarlet itself worth warning for?

fred3m replied:
Not really. See previous comment, with the addendum that this happens quite often for faint sources and there is no need to overload the logger with useless warnings, since 90% of the time the warning would be "this source has a SNR too low to initialize as an `ExtendedSource`.

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+
+def morphToHeavy(source, peakSchema, xy0=Point2I()):
+    """Convert the morphology to a `HeavyFootprint`
+    """
```
Where are the Parameter docs? (2)

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
-    LSST images.
+def checkConvergence(source):
+    """Check that a source converged
     """
```
Where are the Parameter docs? (1)

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+
+
+def modelToHeavy(source, filters, xy0=Point2I(), observation=None, dtype=np.float32):
+    """Convert the model to a `MultibandFootprint`
```
Where are the Parameter docs? (3)

taranu commented on python/lsst/meas/extensions/scarlet/deblend.py:
```diff
@@ -112,21 +108,31 @@ def _computePsfImage(self, position=None):
+
+    Parameters
+    ----------
+    mExposure: `MultibandExposure`
```
See https://developer.lsst.io/python/numpydoc.html#py-docstring-parameters
- Must have a space before and after the colon
- Parameter types should have the full namespace (`lsst.foo.bar`); I haven't checked to see if they work without them
- It would be nice to maintain consistency (always end with a period; only add hyphens for lists)

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,142 @@
+        Record for a peak in the parent `PeakCatalog`
+    observation: `LsstObservation`
+        The images, psfs, etc, of the observed data.
+    bbox: `Rect`
```
I don't think you imported Rect anywhere so this probably won't work.

fred3m replied:
It should, because I don't think this actually creates a link to the class. Otherwise I'd have to import `Rect` but never use it, which would cause `flake8` to fail.

taranu replied:
I thought the whole point was to create links, but apparently it doesn't? I'll have to ask about that.

Anyway, in this case since you're not actually importing Rect, the full namespace would be more useful.

fred3m replied:
Ok. It's a good thing you had me do this, I forgot that the stack doesn't call it `Rect`, it's `Box2I`.

taranu reviewed (changes requested):
Mostly fine but I left a few comments that should be addressed.

Commit:
Update/add docstrings

Commit:
Fix flake8 errors

taranu commented on python/lsst/meas/extensions/scarlet/__init__.py:
```diff
@@ -24,4 +24,4 @@
 from .source import *
 from .blend import *
-from .deblend import *
+from . import deblend
```
Missing from .version import * (the docs won't build without it)

taranu commented on python/lsst/meas/extensions/scarlet/source.py:
```diff
@@ -20,179 +20,197 @@
+
+    Parameters
+    ----------
+    frame: `LsstFrame`
```
Need a space before and after the colon here and in most of the added parameter docstrings.

Sadly I haven't been able to convince PyCharm to do this automatically.

fred3m replied:
This is an odd style guide requirement, as DM typically doesn't adhere to the spaces around operators rule. Numpydocs doesn't require this to compile properly, so this is just a waste of time (not your fault, I'm just venting). But I fixed it throughout the package, since I never use a space on both sides of the colon (because why would a sane person!).

Commit:
Lower detection threshold

Testing has shown that this threshold is too high and results in
truncated boxes around extended objects. We lower the threshold
from 5 to 1 to allow sources to grow larger.

Commit:
Uddate docstrings

fred3m commented:
Should be ready to go now.
